New functionality: Borderlands:TPS, Challenges, Character Name, Others

apocalyptech opened the pull request:
Hello!  Not sure if you'd be interested in merging this at all, but figured I'd pass it along regardless.

I've added in a couple new features.  The big one is supporting editing of Challenge data, so it's possible to (for instance) set all your challenges to one below their "max" value, to more easily bump up your badass ranks or the like.  It's... a bit silly, really, given that there's a Profile editor out there which could be used to just edit your badass ranks directly, but I was struck with the desire to figure out how these were stored in the savegames.  The possibly-more-useful part would be the ability to zero out a character's challenges, which might be useful if someone wanted to reset but wasn't at the required 85% level for an in-game reset.

One potential problem with the changeset is that my code currently assumes little-endian, and I assume it'd produce invalid non-PC savegames (and probably screw up reading non-PC savegames as well).  I didn't see an immediately obvious method of passing around the "endian" var to the necessary functions, so for now it's just hardcoded to assume little.  If you _were_ interested in merging this but wanted that straightened out, and had a non-PC savegame or two lying around, I'd be happy to play around with them (though I wouldn't actually be able to test out the non-PC saves "natively").

Anyway, in addition to the challenge stuff, there's also the ability to set the character's name from the commandline, and also setting the save slot number (which I think is ignored by the engine, actually, but c'est la vie!)

Thanks for the app, regardless, even if you don't want to merge this!  It's always nice to be able to dig into save formats with tools like this.

apocalyptech committed:
Add support for setting challenge data

apocalyptech committed:
Switch to using the internal challenge ID for indexing, since that appears to be more canonical to B2 itself

apocalyptech committed:
Added an option to zero out all non-location-specific challenges

apocalyptech committed:
Add ability to unlock all challenges

apocalyptech committed:
Streamline our challenge unlock, a bit.

apocalyptech committed:
Some changes to how we unlock challenges.  Technically more robust, even if it'll probably never make use of that robustness.  Still, we've also got the official BL2 challenge names associated with the challenges, so there's that.

apocalyptech committed:
Add functionality to set the character name and save_game_id

apocalyptech committed:
Added a little note about zeroing out challenges

apocalyptech committed:
Only set the bonus levels if we haven't reached them yet (or if zero/max have also been specified)

apocalyptech committed:
Option to automatically set the backpack space to max, without having to specify the number

apocalyptech committed:
Replacing optparse with argparse, and switching default endianness to little

apocalyptech committed:
Changing all the former 'modify' options to their own arguments.  Not actually tied-in yet, though.

apocalyptech committed:
Initial version of using our new argparse arguments for modifications.  Totally untested at the moment.

apocalyptech committed:
Pulling basically everything into a big ol' class file.  Very WIP, I'm sure various things got broken.

apocalyptech committed:
Make our debug function a little nicer, and stop passing 'endian' around

apocalyptech committed:
Moving filehandle operations to where they belong, and alter how we deal with our commandline args a bit

apocalyptech committed:
Setting verbose output by default for now

apocalyptech committed:
Don't need those two vars anymore

apocalyptech committed:
A few more changes to how we handle arguments

apocalyptech committed:
Handle exceptions again

apocalyptech committed:
Switching up how our output options work again, and fixing savegame writing so it actually works

apocalyptech committed:
Verified that the current B2 engine doesn't actually support setting skill points.  Removing that option.

apocalyptech committed:
Fixing errors while reading from JSON files

apocalyptech committed:
I suppose it makes sense to link to my own fork here

apocalyptech committed:
Ah, Seraph are crystals, not tokens

apocalyptech committed:
Changing 'unlocks' to 'unlock' on the arguments

apocalyptech committed:
Renaming 'truevaulthunter' to 'tvhm'

apocalyptech committed:
Big ol' update to the README file to support our new argument style

apocalyptech committed:
Make verbose output the default but add a -q/--quiet option to turn it off

apocalyptech committed:
Prettifying output a little bit

apocalyptech committed:
Fixing some wording

apocalyptech committed:
Splitting into a B2 app and a TPS app.  A bit messy in terms of naming conventions, but seems to work all right.  Next up is getting TPS challenge definitions squared away

apocalyptech committed:
Pulling argument parsing into the main class, fixing up TPS item import/export, prepping for black market/ammo changes

apocalyptech committed:
Renaming BL2Error to BorderlandsError since it's no longer BL2-specific

apocalyptech committed:
Clean up the unlock code a bit, and add in ammo unlocking

apocalyptech committed:
Adding --maxammo option, and having ammo unlock automatically do that.

apocalyptech committed:
Some rewording

apocalyptech committed:
Okay, TPS challenge data should be in place now

apocalyptech committed:
Renaming a few things to be less wordy

apocalyptech committed:
Adding in a default for --bank (to max it out automatically)

apocalyptech committed:
Don't default to stdin/stdout, rather require them to be specified

apocalyptech committed:
Notification/verification on file overwrite, and a --force flag to force overwrite

apocalyptech committed:
Getting rid of some ambiguity in arguments - the optional arguments don't play well with positional arguments in some cases, and forcing the arguments seems slightly better to me than having people use -- conditionally

apocalyptech committed:
Fixed a bug in item imports (looks like import was broken in the original for awhile) and adding a bit more verbosity to the import process

apocalyptech committed:
Don't export the empty items apparently present in Borderlands savefiles

apocalyptech committed:
Removing an old option which hasn't done anything in awhile

apocalyptech committed:
Bunch of updates to the README, hopefully will be slightly more readable/useful now

apocalyptech committed:
Renaming 'parsed' back to 'json', and 'json' to 'decodedjson'.  Still slightly confusing, but better, I think.

apocalyptech committed:
A little friendlier help text for --output

apocalyptech committed:
Adding a couple TODO items which have been on my mind, but which I'm unlikely to address any time soon

apocalyptech commented:
Hm, looks like this pull request just updates as I make commits?  Odd.

Anyway, since my original pull request, I've ended up doing a bunch more work on my fork.  Dunno if you'd ever be interested in merging, but here's a basic rundown of what's new:
1. Pulled the code into a class so it's a bit more modular.
2. Added Borderlands: The Pre-Sequel support (which was the reason for the modularity in point 1)
3. Defaults to little-endian (PC) outptut instead of big-endian.
4. Unlocking ammo SDUs / setting ammo levels
5. Fixed an error in the item import code (would generate invalid savegames if there were any weapons in the item file, previously)
6. Added a bunch of stderr output to show what the program's doing
7. Won't overwrite files without either user confirmation, or with a --force option.
8. Switched from optparse to argparse, moved the "-m" option string to individual arguments for each operation

In addition to the stuff I'd mentioned originally, namely setting the character name, save game slot, and the ability to work with Challenge data directly.  There's probably one or two other minor things that I've forgotten about, but that list should be reasonably thorough.

Unfortunately, the code layout's changed enough that providing patches for any ONE of those features is reasonably difficult, so the pull request is probably just gonna be the whole thing.

apocalyptech committed:
Mmf, looks like these currently require Python 2

apocalyptech committed:
Some stricter bounds checking on setting character level, and set XP properly for all character levels (the equation found on the borderlands wikia doesn't appear to be totally accurate at high levels)

apocalyptech committed:
Add a --forceitemlevels flag to set item levels even if they're currently at level 1

apocalyptech committed:
First pass of porting to Python3.  Maybe generally works okay but barely tested

apocalyptech committed:
Add a UVHM unlock.  Wonder why I never got that in here sooner

apocalyptech committed:
Some Python 3 fixes -- I *think* JSON output (and input) should work properly again, though it's not seen *much* testing yet

apocalyptech committed:
A few more Python 3 fixes; exporting items should work okay now

apocalyptech committed:
Some more Python 3 fixes; Borderlands data is latin1, not utf-8

apocalyptech committed:
A few more Python 3 fixes, for dealing with item imports

apocalyptech committed:
A few more Python 3 tweaks

apocalyptech committed:
Fixes for py3 slaughterdome and challenge unlocks.  The sorting in the challenge unlock section is ONLY for our comparison tests w/ the py2 version, not actually needed, but we're leaving it anyway

apocalyptech committed:
Possibly the last of the Python 3 fixes (at least to address broken stuff), should be fully functional on Python 3 now

apocalyptech committed:
A few housekeeping changes for Python 3

apocalyptech committed:
Renaming bltps_save_edit.py to tps_save_edit.py

apocalyptech committed:
Return code if there was an error

apocalyptech committed:
Ah right, explicitly call out to python3 now

xcier reviewed (changes requested):
Hello, I was just wondering if by any chance you could add overpower levels to this I would greatly appreciate it if you could.

apocalyptech commented:
I'm looking into adding that, btw.  It's a weird layout!  Ended up looking into the Gibbed sourcecode, because it wasn't immediately obvious -- turns out that they use some weird little custom items in the item pool to store a few bits of information, including what Overpower levels have been unlocked.  These are the ones which the item export skips and seemed to have all zeroes for the actual item data.  The data in question in this case is all stored in what this project refers to as the item's `unknown2` field (Gibbed calls that one "Quantity").  Anyway, I can at least read the OP level now, should be able to write it out pretty shortly.

Not that I wouldn't still be happy to have my changes merged into the main branch, btw, but if you end up with other feature requests, it may be better to just submit 'em as an issue on my own fork.  :)

(And, since I'm here again - since the last update I'd enumerated above, my fork's also received a port over to Python 3 (Py2 isn't supported over there), so that's something else to keep in mind.)

xcier commented:
Yes I noticed that as well with the -g command above the characters level when I set the level to 80 and was able to use the OP8 gear as there was item tags but no actual data for the OP levels and wasn't sure how to inject it since there is no current listing of it in the .py code glad to see you are still working on this and can't wait to see as I would like to add a GUI to this so easy character creation can exist. such as custom currency or just clicking a tab for max money, skill trees, ammo, etc. Sorry, I am pretty new to github so it took me a while just to find out where I could comment on here.

apocalyptech committed:
Added --oplevel option to set overpower levels (thanks to Gibbed's sourcecode, never would have found that otherwise)

apocalyptech commented:
Okay, I've added an `--oplevel` argument on my fork, which seems to work well!  You can get that from here: https://github.com/apocalyptech/borderlands2/

And no worries about being new to github, it's a complex beast.  :)  Enjoy!

dobey commented:
I've found this through my attempts at trying to convert my PS3 save games over to Steam, on Linux. It seems this is unlikely to get merged (the PR has been open for 3.5 years), as last commit was 6 years ago.  I see you don't have Issues enabled on your fork (which I am trying to use). Any chance you could open up issue reporting on it? I've copied a file over, however, with both the old script, and this fork, I can only get "Invalid save file" as an error. I'd like to open an issue and figure what is going on. Thanks.

apocalyptech commented:
Oh, for sure!  I hadn't even realized that Issues weren't already enabled on there - sorry about that!  Should be good to go now, feel free to submit away over there.  :)

apocalyptech committed:
Only took three months, but finally updating this to support level 80, and OP 10

apocalyptech committed:
Copying mission data, and also enforce max level per game

apocalyptech committed:
Fix an error which could prevent writing to a savefile under some circumstances

apocalyptech committed:
Committing some code that I'd used to organize my savegame archives

apocalyptech committed:
Require Python 3.9, and fix up a function whose definition was changed in Python 3.9.  This closes #3

apocalyptech committed:
Oops, missing a line here which prevented the TPS version from launching

loot-midget committed:
Update savefile.py

New command line option for fixing only challenges that got overflow.

Links to source problem description:

https://steamcommunity.com/app/49520/discussions/0/1327844097129063344/ 

https://steamcommunity.com/app/49520/discussions/0/38596748231645372/

**How to use**
```
$ bl2_save_edit.py -o savegame --fix-challenge-overflow Save0001.sav Save0001_fixed.sav

Opening Save0001.sav for input file

Performing requested changes
fix overflow in: Dolla Dolla Bills, Y'all
fix overflow in: Say "Watt" Again
fix overflow in: Slag-Licked
fix overflow in: Gun Slinger
fix overflow in: Heal Plz

Opening Save0001_fixed.sav for output file
Writing savegame file

Done!
```

apocalyptech committed:
Merge pull request #5 from loot-midget/patch-1

Fix huge negative numbers on challenges

apocalyptech committed:
Add the new arg to the README and slight grammar tweak

loot-midget committed:
New option for printing not fully explored areas.
Should help in World Traveler achievement in Borderlands 2.

--print-unexplored-levels

loot-midget committed:
typo

apocalyptech committed:
Heh, my assumptions about this turned out to be wrong.

apocalyptech committed:
Merge branch 'print_unexplored_levels' of github.com:loot-midget/borderlands2 into loot-midget-print_unexplored_levels

apocalyptech committed:
Report on maps which don't contribute to exploration achievement

apocalyptech committed:
Merge branch 'loot-midget-print_unexplored_levels'

apocalyptech committed:
Fix some improper change detection with --copy-nvhm-missions and split out info-only arg

apocalyptech committed:
Adding a mode for just printing info to the console

apocalyptech committed:
Adding a TOC

apocalyptech committed:
More README tweaks

apocalyptech committed:
Ah yes, I'd gotten to the bottom of the XP thing awhile back

loot-midget committed:
add black formatter (#1)

* add black formatter
* apply black to source files

loot-midget committed:
Fix string formatting (#2)

- remove unnecessary shebang lines
- remove relative imports
- convert all pack/unpack printf formats to str.format()
- convert all other printf formats to f-strings

loot-midget committed:
Common refactoring (#3)

- empty __init__.py
- fix string formatting
- fix struct.pack/unpack prefix
- comment out unused code to reduce refactoring efforts
- cleanup imports
- split savegame to modules
- single launcher code
- fix SDU order
- extract args parsing to config.py
- extract OP check function
- refactor & improve explorer achievements report

loot-midget committed:
use bool() for False/None check (#4)

loot-midget committed:
add actual command output for explorerX option (#5)

apocalyptech committed:
Redirect users to loot-midget's fork

loot-midget committed:
add actual command output for explorerX option (#6)

loot-midget committed:
Challenge Accepted progress (#7)

+ more typing

loot-midget committed:
simplify Challenge Accepted code

loot-midget committed:
- add '_base64' field for raw data of weapons/items
- remove spoiling assert
- rename variable

apocalyptech committed:
Merge remote-tracking branch 'loot-midget/master'

loot-midget committed:
cleanup

loot-midget committed:
extract _set_level

loot-midget committed:
extract _set_money

loot-midget committed:
extract _set_overpowered_level / _set_item_level

loot-midget committed:
reduce cyclic complexity

loot-midget committed:
fix typing

loot-midget committed:
flush streams to avoid mixing stacktrace with general output

loot-midget committed:
- extract unwrap_challenges/wrap_challenges to functions
- get rid of self.config.changes: data is always altered but error will be thrown if there are changes and no output file specified
- add --reset option:
    - reset Bad Touch SMG flag
    - reset Doctor's Orders mission in the highest playthru

loot-midget committed:
add documentation

apocalyptech committed:
Merge remote-tracking branch 'loot-midget/master'
